packages/git/src/browser/git-quick-open-service.ts

```diff
@@ -0,0 +1,134 @@
+    }
+
+    async push(): Promise<void> {
+        const [repository, remotes] = await Promise.all([this.getRepository(), this.getRemotes()]);
+        const execute = (item: QuickOpenItem) => this.git.push(repository, { remote: item.getLabel() });
+        const items = remotes.map(remote => new GitQuickOpenItem(remote, execute));
+        const TODO_currentBranchName = ''; // Otherwise should be the branch name.
+        this.quickOpenService.open(this.getModel(items), this.getOptions(`Pick a remote to push the branch${TODO_currentBranchName} to:`));
+    }
+
+    async pull(): Promise<void> {
+        const [repository, remotes] = await Promise.all([this.getRepository(), this.getRemotes()]);
+        const defaultRemote = remotes[0]; // I wish I could use assignment destructuring here. (GH-413)
+        const executeRemote = async (remoteItem: GitQuickOpenItem<string>) => {
+            // The first remote is the default.
+            if (remoteItem.ref === defaultRemote) {
+                this.git.pull(repository, { remote: remoteItem.getLabel() });
+            } else {
+                // Otherwise we need to propose the branches from
+                const branches = await this.getBranches();
+                const executeBranch = (item: GitQuickOpenItem<Branch>) => this.git.pull(repository, { remote: remoteItem.getLabel() });
+                const toLabel = (branchItem: GitQuickOpenItem<Branch>) => branchItem.ref.name;
+                const branchItems = branches
+                    .filter(branch => branch.type === BranchType.Remote)
+                    .filter(branch => (branch.name || '').startsWith(`${remoteItem.ref}/`))
+                    .map(branch => new GitQuickOpenItem<Branch>(branch, executeBranch, toLabel));
+                this.quickOpenService.open(this.getModel(branchItems), this.getOptions('Select the branch to pull the changes from:'));
+            }
+        };
+        const remoteItems = remotes.map(remote => new GitQuickOpenItem(remote, executeRemote));
+        this.quickOpenService.open(this.getModel(remoteItems), this.getOptions('Pick a remote to pull the branch from:'));
+    }
+
+    async merge(): Promise<void> {
+        const [repository, branches] = await Promise.all([this.getRepository(), this.getBranches()]);
+        const execute = (item: GitQuickOpenItem<Branch>) => this.git.merge(repository, { branch: item.getLabel()! });
+        const toLabel = (item: GitQuickOpenItem<Branch>) => item.ref.name;
+        const items = branches.map(branch => new GitQuickOpenItem<Branch>(branch, execute, toLabel));
+        const TODO_currentBranchName = ''; // Otherwise should be the branch name.
+        this.quickOpenService.open(this.getModel(items), this.getOptions(`Pick a branch to merge into the currently active ${TODO_currentBranchName} branch:`));
+    }
+
+    private getOptions(placeholder: string): QuickOpenOptions.Resolved {
```

Comment:
It should be `QuickOpenOptions`, `Resolved` it is for `QuickOpenService` implementations